### PR TITLE
Add the possibility to open its documentation from an add-on's installation dialog

### LINKED ISSUE #13454: Add ability to read documentation before installing add-ons
### Is your feature request related to a problem? Please describe.
Before installing add-ons, it's possible to read info presented activating the About button, not full add-on help. Though add-on URL can be used to get more info about add-ons, webpages are not always updated, specially when add-ons have multiple version, for example for testing.
### Describe the solution you'd like
The add-on help button should be available before installing an add-on, similar to About button.
### Describe alternatives you've considered
Provide a mechanism to specify documentation for different add-on versions and languages, for example,alist of URLs, though seems tricky and confusing..
### Additional context
Many add-ons include documentation, that maybe updated for different versions and can be more consistent that info available on different websites, where people can publish them.

### PR DESCRIPTION
**Pull request by CyrilleB79**
### Link to issue number:
Fixes #13454

### Summary of the issue:
Feature request: before installing it, the user want to be able to check its documentation.

### Description of user facing changes
* A new button allowing to open the documentation is added in the dialog box asking to confirm the installation of an add-on.
* Also the "About add-on" button label has been shortened to "About"
### Description of development approach
* Extract the documentation from the add-on to a file located in %tmp%\nvdaDocs; the name of the file is a randomly name randomly generated by python's tempfile so that two different files can be opened simultaneously.
* %tmp%\nvdaDocs is cleaned only when NVDA is closed; this way, when the add-on doc file is opened in the browser, pressing F5 allows to reload the page (unless NVDA has been closed/restarted in the meantime).
* The opened documentation is the one in NVDA's language or in English by default.
* If no doc is found (neither in the current language nor in English), the button is greyed out.

### Testing strategy:
Various manual tests:
* With an add-on containing a doc in the current language: pressing the doc button opens documentation containing no-ASCII characters
* With an add-on in English only (no translation): pressing the button opens the English doc
* With an add-on with no documentation: check that the button is greyed out

### Known issues with pull request:
Open questions for reviewers:
1. Should I also add the doc button in the error dialogs appearing when the add-on is too old for current NVDA version or when NVDA version is too old for this add-on? For now I have not done it, but this could be added easily.
2. Is there a better strategy regarding temporary files persistence/cleaning? Just putting the files in %tmp% without cleaning seemd a bit exaggerated to me.

### Change log entries:
New features
`When installing an add-on, you can now read its documentation before installing it. (#13454)`

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

### CODE REVIEW

**Comment by CyrilleB79** [comment]
Cc @nvdaes

**Comment by nvdaes** [comment]
Hi Cyrille:

< 1. Should I also add the doc button in the error dialogs appearing when the add-on is too old for current NVDA version or when NVDA version is too old for this add-on? For now I have not done it, 

I think it should be added, in the same way that the About button is added even when add-ons are incompatible, for consistency, so that documentation can be reviewed and people is encouraged to update NVDA or contact add-on authors to provide a new version of the add-on if needed.

> 2. Is there a better strategy regarding temporary files persistence/cleaning? Just putting the files in %tmp% without cleaning seemd a bit exaggerated to me.

I think it's a good idea to clean folders when NVDA is closed, but imo maybe better to add a different folder for each add-on, so that, if some day NVDA includes the possibility of opening several installation dialogs or in other cases, different documentation files can be opened at the same time for comparison. For instance, maybe different add-ons with similar purposes and users may want to compare the documentation of them to decide which they want to install.

**Comment by XLTechie** [comment]
I would agree with both of Noelia's comments. Always better to design for the 
future, for possibilities that we may not now consider likely, if doing so does 
not add a tun of extra work.

**Comment by CyrilleB79** [comment]
Thanks @nvdaes and @XLTechie  for your comments.

> > 1. Should I also add the doc button in the error dialogs appearing when the add-on is too old for current NVDA version or when NVDA version is too old for this add-on? For now I have not done it,
> 
> I think it should be added, in the same way that the About button is added even when add-ons are incompatible, for consistency, so that documentation can be reviewed and people is encouraged to update NVDA or contact add-on authors to provide a new version of the add-on if needed.

This makes sense. I will do it.

> > 2. Is there a better strategy regarding temporary files persistence/cleaning? Just putting the files in %tmp% without cleaning seemd a bit exaggerated to me.
> 
> I think it's a good idea to clean folders when NVDA is closed, but imo maybe better to add a different folder for each add-on, so that, if some day NVDA includes the possibility of opening several installation dialogs or in other cases, different documentation files can be opened at the same time for comparison. For instance, maybe different add-ons with similar purposes and users may want to compare the documentation of them to decide which they want to install.

With the current version of the code, it's already possible to open two documentation of two different add-ons. You just need to press documentation in the first install dialog before opening the second one. There is no collision since the name of the temp file is randomly generated, e.g. the URL in the address bar of the browser is `file:///C:/Users/Cyrille/AppData/Local/Temp/nvdaDoc/tmpn6dh7h03.html`.

I will update the initial description to clarify this point.

**Comment by nvdaes** [comment]
>  You just need to press documentation in the first install dialog before opening the second one. There is no collision since the name of the temp file is randomly generated, e.g.

Great, thanks.

**Comment by CyrilleB79** [comment]
I have put the "Documentation" button also in error dialogs (see lien aacc100).

Any comment regarding the order of the buttons? For now they are as folllows:
* In confirm dialog: About, Documentation, Yes, No; with About focused first
* In error dialogs: About, Documentation; with About focused first

**Comment by nvdaes** [comment]
> Any comment regarding the order of the buttons?

For me the order is right.

File: source/addonHandler/__init__.py
Comment (nvdaes) [inline source/addonHandler/__init__.py:185]:
```suggestion
def createTempDocDir():

```

File: source/addonHandler/__init__.py
Comment (nvdaes) [inline source/addonHandler/__init__.py:188]: Shouldn't be added try... except to create the dir?

File: source/addonHandler/__init__.py
Comment (nvdaes) [inline source/addonHandler/__init__.py:189]:
```suggestion


```
I think this can be removed, or is there any reason to return the path again, even when there's a function to get it?
For example, in config, initConfigPath, the path is created but not returned.

**Review by nvdaes — CHANGES_REQUESTED** [review]

File: source/addonHandler/__init__.py
Comment (CyrilleB79) [reply source/addonHandler/__init__.py:185]: I am a bit concerned by `createTempDocDir` since the directory is not created if it already exists. And `createTempDocDirIfItDoesNoYetExist` seems a bit long!

File: source/addonHandler/__init__.py
Comment (CyrilleB79) [reply source/addonHandler/__init__.py:189]: Yes, true for current version of the code.

But actually, I am thinking on how to rewrite this code in case the temp "nvdaDoc" directory creation fails for another reason than "already exists" and have not a clear idea on what should be done in this case. Ideas welcome.

**Comment by nvdaes** [comment]
> But actually, I am thinking on how to rewrite this code in case the temp "nvdaDoc" directory creation fails for another reason than "already exists" and have not a clear idea on what should be done in this case.

What about trying to create the directory in the initialize function, and register in the NVDA's log if an exception is raised. In this case, the button to open the documentation would be unavailable for the current session of NVDA.

**Comment by CyrilleB79** [comment]
> > But actually, I am thinking on how to rewrite this code in case the temp "nvdaDoc" directory creation fails for another reason than "already exists" and have not a clear idea on what should be done in this case.
> 
> What about trying to create the directory in the initialize function, and register in the NVDA's log if an exception is raised. In this case, the button to open the documentation would be unavailable for the current session of NVDA.

What you call the initialize function (actually `getTempDocDir`) is run each time you press the button, not only at NVDA startup as it is done for the config folders. I need to check each time that the folder exists (and recreate it if needed), because the user may clear the temp folder content during an NVDA session. I do not find acceptable to disable the doc button until next NVDA restart.

**Comment by nvdaes** [comment]
> I do not find acceptable to disable the doc button until next NVDA restart.

I agree with you after your explanation. Then I think that you should show an error message dialog.

**Comment by CyrilleB79** [comment]
> Could this target the new add-on store in #13985 instead?

I will look at it if you wish.

However, is there an ETA for the add-on store? It has been announced for years and has not yet been released. With Reef leaving NVAccess, I wonder when this feature will become a reality (i.e. be merged in alpha). If it is not expected in 2023, I would prefer to have this PR already merged in current NVDA's codebase. Just my opinion though.

**Comment by seanbudd** [comment]
Work on the add-on store has resumed. Unless there are further unexpected delays (like a slew of security issues) it should make it into 2023.

**Comment by seanbudd** [comment]
I think this might be better left as blocked until the new add-on store is merged to master.

**Comment by CyrilleB79** [comment]
> I think this might be better left as blocked until the new add-on store is merged to master.


OK.

**Comment by seanbudd** [comment]
This is now unblocked 🥳

**Comment by amirmahdifard** [comment]
@CyrilleB79 hi, i want to request something from you if possible. I don't want you to update this to the current codebace or anything because i don't wish to try this with the last addon store, please just do something so that the ap veyor completes the build again, i want to try the snappshot of this pr. Is it possible for you to do this? because ap veyor doesn't have the build of this, as it is deleted after a month. Thanks

**Comment by CyrilleB79** [comment]
@amirmahdifard, I have tried to push an empty commit to re-trigger a build. Unfortunately, appVeyor does not rebuild, maybe due to merge conflicts (even if I have not merged from master). You'll need to build yourself, sorry.

**Comment by amirmahdifard** [comment]
@CyrilleB79 i have an idea. please, instead of empty commit, for example, update a line of text in the documemtations, or something, so it will build may be. If that also doesn't work, i won't annoy you anymore. Thanks.

**Comment by amirmahdifard** [comment]
i sadly don't know how to build nvda, and that's my only reason why i don't contribute to nvda's code, but i contribute to addons. I know that asking you to build for me is a bit rood, so when ap veyor build fales this time, i will just give up

**Comment by CyrilleB79** [comment]
The empty commit is not the cause; I have already used them to retrigger builds.

If you want to learn how to build NVDA, you can read the documentation to [create your development environment](https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/createDevEnvironment.md) or more generally the [contributing guide](https://github.com/nvaccess/nvda/blob/master/.github/CONTRIBUTING.md). And if you have questions, subscribe to [nvda-devel](https://groups.io/g/nvda-devel/topics) mailing list and ask there.

**Comment by amirmahdifard** [comment]
@CyrilleB79 ok, Thanks for your try. Sadly, i don't see my self advance enough to be able to build nvda. But if i could, i would implement a lot of cool things for nvda, as i know how to write code but i don't know how to build nvda. Cn you tell me how can i setup ap veyor on one of my github repos? do i need to modify my code? or how the ap veyor can build? Thanks, and i know here is not a right place to talk about this probably

**Comment by hwf1324** [comment]
> @CyrilleB79 ok, Thanks for your try. Sadly, i don't see my self advance enough to be able to build nvda. But if i could, i would implement a lot of cool things for nvda, as i know how to write code but i don't know how to build nvda. Cn you tell me how can i setup ap veyor on one of my github repos? do i need to modify my code? or how the ap veyor can build? Thanks, and i know here is not a right place to talk about this probably

You don't need to use AppVeyor to build NVDA, the link above already gives the process of building NVDA locally.
Unless your local computer is not readily available.

I'm sorry I don't have experience building projects with AppVeyor, you'll need to check the AppVeyor documentation. I haven't found a personal process for building NVDA using AppVeyor in the repository, and NVDA has some of its own AppVeyor scripts, not sure it works out of the box.

**Comment by seanbudd** [comment]
Yes there needs to be a clean merge to the latest master for AppVeyor to build, as any PR is tested with the code merged into the target branch. I'd ask that any discussion around building NVDA be moved to a separate GitHub discussion outside of this PR.

**Comment by seanbudd** [comment]
Should we close this PR as abandoned?

**Pull request closed without merging**
